subsys/power/device.c

```diff
@@ -99,51 +112,58 @@ int sys_pm_force_suspend_devices(void)
 
 void sys_pm_resume_devices(void)
 {
-	int i;
+	device_idx_t pmi = num_pm - num_susp;
 
-	for (i = 0; i < device_count; i++) {
-		if (!device_retval[i]) {
-			int idx = device_ordered_list[i];
+	num_susp = 0;
+	while (pmi < num_pm) {
+		device_idx_t idx = pm_devices[pmi];
 
-			device_set_power_state(&pm_device_list[idx],
-					DEVICE_PM_ACTIVE_STATE, NULL, NULL);
-		}
+		device_set_power_state(&all_devices[idx],
+				       DEVICE_PM_ACTIVE_STATE,
+				       NULL, NULL);
+		++pmi;
 	}
 }
 
 void sys_pm_create_device_list(void)
 {
 	int count;
-	int i, j;
-	bool is_core_dev;
+	device_idx_t pmi;
 
 	/*
 	 * Create an ordered list of devices that will be suspended.
 	 * Ordering should be done based on dependencies. Devices
 	 * in the beginning of the list will be resumed first.
 	 */
-	device_list_get(&pm_device_list, &count);
+	device_list_get(&all_devices, &count);
 
-	/* Reserve for 32KHz, 16MHz, system clock, etc... */
-	device_count = ARRAY_SIZE(core_devices);
+	__ASSERT_NO_MSG(count <= (device_idx_t)(-1));
 
-	for (i = 0; (i < count) && (device_count < MAX_PM_DEVICES); i++) {
+	/* Reserve initial slots for core devices. */
+	num_pm = ARRAY_SIZE(core_devices);
+
+	for (pmi = 0; (pmi < count) && (num_pm < MAX_PM_DEVICES); pmi++) {
+		device_idx_t cdi = 0;
+		const struct device *dev = &all_devices[pmi];
+
+		/* Ignore "device"s that don't support PM */
+		if (dev->config->device_pm_control == device_pm_control_nop) {
+			continue;
+		}
 
-		/* Check if the device is core device */
-		for (j = 0, is_core_dev = false;
-		     j < ARRAY_SIZE(core_devices);
-		     j++) {
-			if (!strcmp(pm_device_list[i].config->name,
-						core_devices[j])) {
-				is_core_dev = true;
+		/* Check if the device is a core device, which has a
+		 * reserved slot */
+		while (cdi < ARRAY_SIZE(core_devices)) {
+			if (strcmp(dev->config->name, core_devices[cdi]) == 0) {
+				pm_devices[cdi] = pmi;
 				break;
 			}
+			++cdi;
```

Comment:
is there a reason why you use this syntax here?

Reply:
Do you mean `++cdi` instead of `cdi++`?  Yes, it's my automatic choice for cases where I don't care about the value of the expression.  The compiler will probably discard the temporary anyway but it might improve code generation.

Reply:
> but it might improve code generation
Could detail a bit more on this? I fail to see how it could happen actually. If the temporary one is discarded and only the incremented one is kept, isn't the generated code be the exactly the same?

Reply:
In C++ `type& operator++()` (for `++e`) and `type operator++(int)` (for `e++`) need to be implemented for user-defined types, and the latter is generally more complex because the original value has to be preserved during the calculation and update of the incremented stored value.  Thus in a loop like:

    for (auto it = begin() ; it != end(); ++it) {}

you definitely don't want to use `it++` because the compiler's forced to use an implementation that may be more complex just to create a temporary it will throw away.

I use it regardless of language for consistency and clarity of intent.  Here, I don't care about the old value, so I use the expression that doesn't produce it.

Reply:
ok, by code generation you meant the in-between code before the actual optimizations and binary generation. I thought you meant the binary one.